uplc.md

```diff
@@ -0,0 +1,205 @@
+
+Here is a simple example of a UPLC program that is equivalent to the Haskell `id` function:
+
+```lisp
+Program
+  ()
+  (Version () 1 0 0)
+  (LamAbs
+     ()
+     (DeBruijn {dbnIndex = 0})
+     (Var () (DeBruijn {dbnIndex = 1})))
+```
+
+## Pluto
+
+As you might have noticed it is not practical to write UPLC by hand. [Pluto](https://github.com/Plutonomicon/pluto) is a simple programming language that assembles directly to UPLC; that is to say, it 'maps' directly (more or less) to the AST of UPLC. Understanding Pluto facilitates an understanding of UPLC. Here's the equivalent Pluto program for the above UPLC example:
+
+```haskell
+(\× -> x)
+```
+
+Pluto retains the structure of UPLC, and enables you to write UPLC "by hand" (as it were) but in a more ergonomic way.
+
+## UPLC Representation
+
+Plutus represents a UPLC program using the following Haskell type:
+
+```haskell
+data Program name uni fun ann = 
+  Program ann (TPLC.Version ann) (Term name uni fun ann)
+```
+
+The part we are interested in is `Term name uni fun ann`. A UPLC is just a `Term` defined recursively in terms of itself, similar to languages like Lisp. The `Term` type is defined as follows:
+
+```haskell
+data Term name uni fun ann
+    = Var !ann !name
+    | LamAbs !ann !name !(Term name uni fun ann)
+    | Apply !ann !(Term name uni fun ann) !(Term name uni fun ann)
+    | Force !ann !(Term name uni fun ann)
+    | Delay !ann !(Term name uni fun ann)
+    | Constant !ann !(Some (ValueOf uni))
+    | Builtin !ann !fun
+    | Error !ann
+```
+
+`Term` is parametrized over two type variables of particular interest to us.
+
+1. `fun` is the list of *builtin functions* supported by `Term` or `Program` (see UPLC builtins below), and 
+2. `uni` is the kind of *value types* supported by `Term` or `Program`. 
+
+They are normally instantiated to `DefaultFun` and `DefaultUni` respectively, and Pluto uses exactly these.
+
+## UPLC builtins
+
+The *builtin functions* supported by UPLC are specified by the `DefaultFun` type. The Plutus evaluator (which all cardano nodes in the blockchain use to execute script validators) provides the implementation for them, and they are available in the same module (`PlutusCore.Default.Builtins`).  Here is a subset of those builtins:
+
+```haskell
+data DefaultFun
+    = AddInteger
+    | SubtractInteger
+    ...
+    | AppendByteString
+    | ConsByteString    
+    ...
+    | Sha2_256
+    | VerifySignature
+    ...
+    | IfThenElse
+    ...
+```
+
+Bear in mind that UPLC is untyped, therefore these builtin functions will fail to evaluate if their types don't match what they expect.
+
+## UPLC values
+
+The *values* supported by UPLC are specified by the `DefaultUni` type.
+
+```haskell
+data DefaultUni a where
```

Comment:
I think you should include these guys in this section as well https://github.com/input-output-hk/plutus/blob/e995df9a339b69523e34bad35816ee1e4ddd9669/plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs#L91-L94